### ISSUE TITLE
pass pair plot scatter_kwargs to plt.scatter(), not plt.plot()

### ISSUE THREAD

**Issue by grahamgower** [arviz-devs/arviz#1889]
I want to do a pair plot with a colour map, where the colour of the points indicates the posterior probability. I expected to just be able to pass `scatter_kwargs={"c": probs}` to arviz.plot_pair(), but it seems the matplotlib backend uses plt.plot() instead of plt.scatter(), and plt.plot() doesn't seem to support this (I think?). Also, it's kinda confusing that this is named `scatter_kwargs`, when it doesn't use the scatter function.

**Comment by tattvams** [comment]
Hey can you please elaborate this problem and what exactly is that you are trying to achieve and can you post  any screenshots of the problem.

**Comment by OriolAbril** [comment]
iiuc, the main goal is being able to use the `c` argument in `scatter` which is similar to `color` in `plot` but a bit more flexible. 

That would mean updating the docs at https://github.com/arviz-devs/arviz/blob/9d74000d3d0a1599b017a5dc612388a0663d2f36/arviz/plots/pairplot.py#L100, the [dealiasing actually assumes this is passed down to `scatter`](https://github.com/arviz-devs/arviz/blob/9d74000d3d0a1599b017a5dc612388a0663d2f36/arviz/plots/backends/matplotlib/pairplot.py#L50) so bonus points for fixing a bug we didn't know we had :sweat_smile: and then it should be passed to `ax.scatter` instead of `ax.plot` in https://github.com/arviz-devs/arviz/blob/main/arviz/plots/backends/matplotlib/pairplot.py#L177 and https://github.com/arviz-devs/arviz/blob/main/arviz/plots/backends/matplotlib/pairplot.py#L177

**Comment by tattvams** [comment]
Hey, so I looked a bit into this and `c` argument in `scatter` is working for example in this code : 
`import matplotlib.pyplot as plt

import arviz as az

az.style.use("arviz-darkgrid")

centered = az.load_arviz_data("centered_eight")

coords = {"school": ["Choate", "Deerfield"]}
colors = [0.3,0.8,0.7]
az.plot_pair(
    centered, var_names=["theta", "mu", "tau"], coords=coords, divergences=True, textsize=22, scatter_kwargs={"c": colors}
)
plt.show()`
![image](https://user-images.githubusercontent.com/73391843/156199185-4dceb9f7-39dd-4530-90e0-d1dfdd7800b7.png)
And the reason why dealisaing is assuming that this is passed down to scatter is because `scatter_kwargs` returns kind `scatter`
which i am assuming should be the same (correct me if I am wrong and if i should change it). I did commit some changes for ax.plot -> ax.scatter. Let me know if I understood something wrong and should change it

**Comment by OriolAbril** [comment]
The scatter_kwargs with a `c` key works after your changes or does so already in the current code?

> And the reason why dealisaing is assuming that this is passed down to scatter is because scatter_kwargs returns kind scatter
which i am assuming should be the same (correct me if I am wrong and if i should change it).

The dealiasing should not be changed. It will be correct once the issue is fixed, but is currently incorrect because it uses "scatter" as the second argument instead of "plot", but those arguments are _incorrectly_ passed to an `ax.plot` instance right now.

**Comment by tattvams** [comment]
It worked without my changes and isn't dealiasing using scatter keyword only when  `scatter_kwargs `.... So is there no other keyword like `plot_kwargs` ?

**Comment by grahamgower** [comment]
Hi @tattvams,

It doesn't currently work. Let me illustrate. The following code generates random `x` and `y` coordinates, and also some `z` value based on the `x` and `y` coordinates. Matplotlib's `scatter()` function can plot the value of this `z` variable using a colour map by passing it in the `c=` parameter. In the plot, each dot has a different colour, according to the specific `z` value.
```python
import matplotlib.pyplot as plt
import numpy as np

rng = np.random.default_rng(1234)
x, y, z = rng.uniform(0, 10, size=(3, 100))
plt.scatter(x, y, c=z)
plt.show()
```
![Figure_1](https://user-images.githubusercontent.com/8665077/156390890-54f75fb9-da70-4c2d-825b-70d90a3f3969.png)


If you change the `plt.scatter` to instead be `plt.plot`, you will recieve the following error message.
```
Traceback (most recent call last):
  File "/tmp/scatter.py", line 7, in <module>
    plt.plot(x, y, c=z)
  File "/usr/lib/python3.10/site-packages/matplotlib/pyplot.py", line 2757, in plot
    return gca().plot(
  File "/usr/lib/python3.10/site-packages/matplotlib/axes/_axes.py", line 1632, in plot
    lines = [*self._get_lines(*args, data=data, **kwargs)]
  File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 312, in __call__
    yield from self._plot_args(this, kwargs)
  File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 538, in _plot_args
    return [l[0] for l in result]
  File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 538, in <listcomp>
    return [l[0] for l in result]
  File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 531, in <genexpr>
    result = (make_artist(x[:, j % ncx], y[:, j % ncy], kw,
  File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 351, in _makeline
    seg = mlines.Line2D(x, y, **kw)
  File "/usr/lib/python3.10/site-packages/matplotlib/lines.py", line 370, in __init__
    self.set_color(color)
  File "/usr/lib/python3.10/site-packages/matplotlib/lines.py", line 1030, in set_color
    mcolors._check_color_like(color=color)
  File "/usr/lib/python3.10/site-packages/matplotlib/colors.py", line 130, in _check_color_like
    raise ValueError(f"{v!r} is not a valid value for {k}")
ValueError: array([12.74539857,  9.99356751, 12.25003663,  4.36538875,  8.14602254,
        8.78661633,  9.2902787 ,  5.66688708, 15.8612584 ,  5.79305495,
        7.18222539, 15.3973602 , 11.43216566, 18.12592686,  8.76898665,
       12.25352763, 11.16060829, 10.30860385,  2.99619853, 18.21754524,
        7.88966015,  8.59233907, 13.63997113,  7.03851005,  6.16538883,
       10.80263408,  4.82644868,  9.25285528,  6.94186349,  9.78244537,
       15.27522036, 10.16853085, 16.14819546, 14.5970087 ,  2.51296569,
       11.33555053, 11.44611872, 17.38216807,  0.93636827, 14.14022259,
       14.84411785, 11.91736832, 10.21676171, 15.65048893, 13.45479257,
        6.22682582,  4.33358728, 11.47217771,  1.24086803,  4.64838705,
       10.5199589 , 13.35589471,  7.31870321, 11.48630773, 18.42065178,
       14.74225064, 13.98589875,  5.16273114,  7.96485267,  9.7928    ,
        4.15582   , 12.64464648, 17.21968744, 10.69554244, 10.16331922,
       13.92454338,  9.35741315,  4.71664657,  9.48998161,  8.4562641 ,
       10.18939222,  0.98601694,  7.90308634,  8.90914178, 13.07089323,
        0.67881647, 10.5965448 ,  8.31221063, 13.3089983 ,  7.58058182,
        9.43381122,  9.22571312, 12.49790034, 11.15509301, 10.46866392,
        8.63603748, 12.86335841,  7.78792664, 11.90037994, 10.41934229,
       13.32457332, 10.19184625,  7.07384512,  0.42243579,  6.8327167 ,
       10.90716744,  5.70878519, 15.64657765, 18.88753084,  3.33478009]) is not a valid value for color
```

Your example above with `colors = [0.3,0.8,0.7]` doesn't produce an error, because `[0.3, 0.8, 0.7]` are interpreted as RGB values. In your example, all the points are given that one colour.

**Comment by tattvams** [comment]
> Hi @tattvams,
> 
> It doesn't currently work. Let me illustrate. The following code generates random `x` and `y` coordinates, and also some `z` value based on the `x` and `y` coordinates. Matplotlib's `scatter()` function can plot the value of this `z` variable using a colour map by passing it in the `c=` parameter. In the plot, each dot has a different colour, according to the specific `z` value.
> 
> ```python
> import matplotlib.pyplot as plt
> import numpy as np
> 
> rng = np.random.default_rng(1234)
> x, y = rng.uniform(0, 10, size=(2, 100))
> z = x + y
> plt.scatter(x, y, c=z)
> plt.show()
> ```
> 
> ![Figure_1](https://user-images.githubusercontent.com/8665077/156390890-54f75fb9-da70-4c2d-825b-70d90a3f3969.png)
> 
> If you change the `plt.scatter` to instead be `plt.plot`, you will recieve the following error message.
> 
> ```
> Traceback (most recent call last):
>   File "/tmp/scatter.py", line 7, in <module>
>     plt.plot(x, y, c=z)
>   File "/usr/lib/python3.10/site-packages/matplotlib/pyplot.py", line 2757, in plot
>     return gca().plot(
>   File "/usr/lib/python3.10/site-packages/matplotlib/axes/_axes.py", line 1632, in plot
>     lines = [*self._get_lines(*args, data=data, **kwargs)]
>   File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 312, in __call__
>     yield from self._plot_args(this, kwargs)
>   File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 538, in _plot_args
>     return [l[0] for l in result]
>   File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 538, in <listcomp>
>     return [l[0] for l in result]
>   File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 531, in <genexpr>
>     result = (make_artist(x[:, j % ncx], y[:, j % ncy], kw,
>   File "/usr/lib/python3.10/site-packages/matplotlib/axes/_base.py", line 351, in _makeline
>     seg = mlines.Line2D(x, y, **kw)
>   File "/usr/lib/python3.10/site-packages/matplotlib/lines.py", line 370, in __init__
>     self.set_color(color)
>   File "/usr/lib/python3.10/site-packages/matplotlib/lines.py", line 1030, in set_color
>     mcolors._check_color_like(color=color)
>   File "/usr/lib/python3.10/site-packages/matplotlib/colors.py", line 130, in _check_color_like
>     raise ValueError(f"{v!r} is not a valid value for {k}")
> ValueError: array([12.74539857,  9.99356751, 12.25003663,  4.36538875,  8.14602254,
>         8.78661633,  9.2902787 ,  5.66688708, 15.8612584 ,  5.79305495,
>         7.18222539, 15.3973602 , 11.43216566, 18.12592686,  8.76898665,
>        12.25352763, 11.16060829, 10.30860385,  2.99619853, 18.21754524,
>         7.88966015,  8.59233907, 13.63997113,  7.03851005,  6.16538883,
>        10.80263408,  4.82644868,  9.25285528,  6.94186349,  9.78244537,
>        15.27522036, 10.16853085, 16.14819546, 14.5970087 ,  2.51296569,
>        11.33555053, 11.44611872, 17.38216807,  0.93636827, 14.14022259,
>        14.84411785, 11.91736832, 10.21676171, 15.65048893, 13.45479257,
>         6.22682582,  4.33358728, 11.47217771,  1.24086803,  4.64838705,
>        10.5199589 , 13.35589471,  7.31870321, 11.48630773, 18.42065178,
>        14.74225064, 13.98589875,  5.16273114,  7.96485267,  9.7928    ,
>         4.15582   , 12.64464648, 17.21968744, 10.69554244, 10.16331922,
>        13.92454338,  9.35741315,  4.71664657,  9.48998161,  8.4562641 ,
>        10.18939222,  0.98601694,  7.90308634,  8.90914178, 13.07089323,
>         0.67881647, 10.5965448 ,  8.31221063, 13.3089983 ,  7.58058182,
>         9.43381122,  9.22571312, 12.49790034, 11.15509301, 10.46866392,
>         8.63603748, 12.86335841,  7.78792664, 11.90037994, 10.41934229,
>        13.32457332, 10.19184625,  7.07384512,  0.42243579,  6.8327167 ,
>        10.90716744,  5.70878519, 15.64657765, 18.88753084,  3.33478009]) is not a valid value for color
> ```
> 
> Your example above with `colors = [0.3,0.8,0.7]` doesn't produce an error, because `[0.3, 0.8, 0.7]` are interpreted as RGB values. In your example, all the points are given that one colour.

Oh ok, I guess then there should be a problem with matplotlib backend because in "c" keyword is used and then passed to scatter which returns [PathCollection](https://matplotlib.org/3.5.1/api/collections_api.html#matplotlib.collections.PathCollection)

**Comment by tattvams** [comment]
hey, by backend I didn't mean matplotlib what i meant was maybe there was a reason that c didn't accept any values other than rgb because it may have failed to add cmap functionality which i understand is not the case and that it went to plt.plot() instead of plt.scatter() and can you send me source code where .plot() function is defined because I am unable to find it. Thanks!

**Comment by OriolAbril** [comment]
Do you mean this https://matplotlib.org/stable/api/_as_gen/matplotlib.axes.Axes.plot.html?

**Comment by ahartikainen** [comment]
I think originally we went with plot interface was due to speed (scatter was slower for a huge number of points).

I don't know what is the current standing, but I think going with .scatter solution is fine.


Here is source for .scatter

https://github.com/matplotlib/matplotlib/blob/f33f5ab12142bcef39e1f8cf5c8fb17d5cce0057/lib/matplotlib/axes/_axes.py#L4278

and here source for .plot

https://github.com/matplotlib/matplotlib/blob/f33f5ab12142bcef39e1f8cf5c8fb17d5cce0057/lib/matplotlib/axes/_axes.py#L1396

**Comment by grahamgower** [comment]
Maybe setting `rasterize=True` improves the speed when there are many points?

The main reason I didn't immediately send a patch fixing this in arviz is because I wasn't sure if there were backwards-compatibilitiy concerns over such a change. Figure size/speed issues would also be a bad behaviour change.

**Comment by ahartikainen** [comment]
We could probably add that as an default.

Tbh I think scatter is fast enough these days. Speed is always an issue when handling large datasets, but in most cases the datasets are not that large.

**Comment by OriolAbril** [comment]
closed by #2069
